### PR TITLE
close motion before validation if necessary

### LINKED ISSUE #3203: Unable to set proposal outcome
I click 'set outcome' on a recently closed proposal and the modal pops up. I type in the message and click 'publish outcome'. The 'loading' circle spins for a moment, but nothing happens. I press 'publish outcome again' - same thing. I tried on Chrome and Firefox - same thing.

### PR DESCRIPTION
**Pull request by robguthrie**
fixes #3203

### CODE REVIEW

File: app/models/motion.rb [hunk 1]
```diff
@@ -96,6 +98,18 @@ def closed?
...
 
+  def close_if_lapsed!
+    close! if (!closed? and closing_at < Time.now)
+  end
+
+  def close!
+    did_not_votes.delete_all
+    non_voters = group_members - voters
+    DidNotVote.create! non_voters.map { |user| {motion: self, user: user} }
+    update(closed_at: Time.now,
```
Comment (gdpelican) [inline app/models/motion.rb]: Do we need Time.zone.now here?

File: app/models/motion.rb [hunk 2]
```diff
@@ -44,6 +44,8 @@ class Motion < ActiveRecord::Base
   has_paper_trail only: [:name, :description, :closing_at, :outcome]
 
   after_initialize :set_default_closing_at
+  before_validation :close_if_lapsed!
```
Comment (gdpelican) [inline app/models/motion.rb]: I think this is totally okay to go in, but want to keep my ear out for a better way. Here's what makes me a little uncomfortable:
- I think it's a little odd to have such a slow event (load up a bunch of relations, create a bunch of models, then save = lots of stuff for the user to wait on) happen before validation.
- 'before validation' happens quite a lot, esp. in the test suite. Also, we'd get slightly weird behaviour if a motion is invalid, tries to save, but still closes. ('I'm updating the title to blank; whoops that didn't work. Wait why is it closed now, and I can't edit anymore?')

I also wonder if it'd be possible to bump the DidNotVotes creation (which is the slow bit) off to a side job after the motion has closed?

```
EventBus.listen('motion_closed', 'motion_closed_by_user') do |motion|
  motion.did_not_votes.delete_all
  motion.did_not_votes.create!((motion.group_members - motion.voters).map { |user| {motion: motion, user: user} }
end
```

💛

**Comment by robguthrie** [comment]
Thanks @gdpelican. I'm not happy about it either.

Infact it's a buggy solution because when it closes itself it won't create events. So I don't think we should merge it yet.

The core of the problem is there is always going to be some amount of time where "lapsed" motions exist. But I think I got carried away deciding how bad the problem is - maybe we should just close_if_lapsed as a step in the updating methods of the motion service?

File: app/models/motion.rb (hunk 1 above)
Comment (robguthrie) [inline app/models/motion.rb]: Nope: Time.now will return UTC time, which is fine for the db.

File: spec/services/motion_service_spec.rb [hunk 3]
```diff
@@ -35,7 +35,7 @@
         MotionService.create(motion: motion, actor: user)
       end
 
-      it "syncs the discussion's search vector" do
+      it "syncs the discussion's search vector", focus: true do
```
Comment (gdpelican) [inline spec/services/motion_service_spec.rb]: Will want to drop the `focus` here before merge

File: app/models/motion.rb [hunk 4]
```diff
@@ -96,6 +97,18 @@ def closed?
     closed_at.present?
   end
 
+  def needs_to_be_closed?
+    (!closed? and closing_at < Time.now)
```
Comment (gdpelican) [inline app/models/motion.rb]: Yeah so much better.

**Comment by robguthrie** [comment]
@gdpelican I reckon this is good to merge. Don't want to background the did_not_votes because I'm not sure we know how that will behave in the system.

Good news is we're improving the perf of it by moving from many inserts to one bulk insert.

File: spec/services/motion_service_spec.rb (hunk 3 above)
Comment (robguthrie) [inline spec/services/motion_service_spec.rb]: done ta

File: app/services/motion_service.rb [hunk 5]
```diff
@@ -36,27 +37,20 @@ def self.reopen(motion, close_at)
...
-    motion.closed_at = Time.now
-    motion.save!
-
+    motion.close!
     EventBus.broadcast('motion_close_by_user', motion, user)
     Events::MotionClosedByUser.publish!(motion, user)
   end
 
   def self.create_outcome(motion:, params:, actor:)
+    close(motion) if motion.needs_to_be_closed?
```
Comment (gdpelican) [inline app/services/motion_service.rb]: I found myself thinking 'is it odd that an unauthorized user can close a lapsed motion by trying to create an outcome for it?'

But... meh. 💛

**Comment by gdpelican** [comment]
This is great! 🍏

**Pull request merged**